### PR TITLE
[release/7.0][wasm-mt] Create a Microsoft.NET.Runtime.Emscripten.Cache

### PR DESCRIPTION
**Pull request by lambdageek**
Spin out the emscripten system libraries cache into a separate nupkg.

- Split up the Emscripten SDK by moving the system libraries cache into a separate nupkg.

- make a new Emscripten Cache SDK that includes just the cache directory

- the Cache SDK overrides the WasmCachePath (if it's not set by the user) to point into the Cache SDK

- we leave the mt multithreaded libraries in the cache.  (But we remove the ww WebWorker (non-pthread) Emscripten API libraries)

Addresses dotnet/runtime#75263

### CODE REVIEW

**Comment by lambdageek** [comment]
<details>
<summary>
<pre>
% unzip -l artifacts/packages/Release/Shipping/Microsoft.NET.Runtime.Emscripten.3.1.12.Threading.Sdk.osx-x64.7.0.0-dev.nupkg
</pre>
</summary>
<pre>
Archive:  artifacts/packages/Release/Shipping/Microsoft.NET.Runtime.Emscripten.3.1.12.Threading.Sdk.osx-x64.7.0.0-dev.nupkg
  Length      Date    Time    Name
---------  ---------- -----   ----
      551  09-09-2022 09:34   _rels/.rels
     1051  09-09-2022 09:34   Microsoft.NET.Runtime.Emscripten.3.1.12.Threading.Sdk.osx-x64.nuspec
     1326  09-08-2022 20:20   LICENSE
       63  09-09-2022 13:34   data/UnixFilePermissions.xml
     7006  08-19-2022 00:48   Icon.png
      586  09-08-2022 21:02   Sdk/Sdk.props
      372  09-09-2022 13:24   Sdk/Sdk.targets
  1676538  05-27-2022 00:07   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libasan_rt-mt.a
  6587186  05-27-2022 00:00   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libc++-mt.a
   936608  05-26-2022 23:57   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libc++abi-mt.a
  3524818  05-26-2022 23:40   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a
   539898  05-26-2022 23:04   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libcompiler_rt-mt.a
   543568  05-26-2022 23:05   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libcompiler_rt-wasm-sjlj-mt.a
    81206  05-27-2022 00:02   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libdlmalloc-mt.a
    69552  05-27-2022 00:02   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libemmalloc-memvalidate-mt.a
    77874  05-27-2022 00:02   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libemmalloc-memvalidate-verbose-mt.a
    51398  05-27-2022 00:02   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libemmalloc-mt.a
    77396  05-27-2022 00:02   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libemmalloc-verbose-mt.a
    13750  05-27-2022 00:03   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libfetch-mt.a
    39332  05-27-2022 00:03   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libGL-mt.a
   221616  05-27-2022 00:06   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/liblsan_common_rt-mt.a
   506594  05-27-2022 00:07   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/liblsan_rt-mt.a
    33166  05-26-2022 23:57   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libprintf_long_double-mt.a
 10612496  05-27-2022 00:06   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libsanitizer_common_rt-mt.a
    40380  05-26-2022 23:57   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libsockets-mt.a
    33822  05-26-2022 23:57   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libsockets_proxy-mt.a
    30762  05-27-2022 00:06   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libubsan_minimal_rt-mt.a
   616800  05-27-2022 00:06   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libubsan_rt-mt.a
      400  05-27-2022 00:01   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libunwind-mt.a
  6458328  05-27-2022 00:06   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libwasmfs-asan-debug-mt.a
  6395750  05-27-2022 00:05   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libwasmfs-asan-mt.a
  5270808  05-27-2022 00:04   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libwasmfs-debug-mt.a
  5238554  05-27-2022 00:04   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libwasmfs-mt.a
   187344  05-27-2022 00:03   tools/emscripten/cache/sysroot/lib/wasm32-emscripten/libwebgpu_cpp-mt.a
      778  09-09-2022 09:34   [Content_Types].xml
      695  09-09-2022 09:34   package/services/metadata/core-properties/31e400e0c66f4840bdcd0dad77f33a82.psmdcp
</pre>
</details>

**Comment by lambdageek** [comment]
Using `-L` doesn't work because emscripten is very particular about how it looks for libraries that it considers "system" libs.  There's a python script that isn't very flexible.

But, we can change the directory where the cache resides.

And locally, if I exclude the `*-ww-*.a` libraries (since we're not using the non-pthreads Emscripten WebWorkers APIs), the cache dir is only 173 megs.

So instead of making a "Emsciprten.Threading.Sdk" package.  we could make an "Emscripten.Cache.Sdk" package instead and put all the libs (threaded and non-threaded) in there, and use the nice [`WasmCachePath` SDK property](https://github.com/dotnet/runtime/blob/eaa86a0d3e318a20dfc4bd8735ac9332cd0d8d06/src/mono/wasm/build/WasmApp.Native.targets#L262) to override the `EM_CACHE` environment variable that we set when doing the linking.

This appears to work locally:

```
  <PropertyGroup>
    <EmscriptenThreadingSdkCacheDir>/home/aleksey/working/dotnet-runtime/src/mono/wasm/emsdk/upstream/emscripten/cache/</EmscriptenThreadingSdkCacheDir>
  </PropertyGroup>

  <Target Name="_EmscriptenThreadingPrependLinkFlags" BeforeTargets="_PrepareForWasmBuildNative" Condition="'$(WasmEnableThreads)' == 'true'">
    <PropertyGroup>
      <WasmCachePath>$(EmscriptenThreadingSdkCacheDir)</WasmCachePath>
    </PropertyGroup>
  </Target>
```

**Comment by lewing** [comment]
@joeloff if emsdk adds a new pack is that going to cause you issues?

**Comment by lambdageek** [comment]
```
$ ls -lh artifacts/packages/Release/Shipping/
total 271M
-rw-r--r-- 1 aleksey aleksey  50M Sep  9 13:05 Microsoft.NET.Runtime.Emscripten.3.1.12.Cache.Sdk.linux-x64.7.0.0-dev.nupkg
-rw-r--r-- 1 aleksey aleksey  25M Sep  9 13:04 Microsoft.NET.Runtime.Emscripten.3.1.12.Node.linux-x64.7.0.0-dev.nupkg
-rw-r--r-- 1 aleksey aleksey 196M Sep  9 13:04 Microsoft.NET.Runtime.Emscripten.3.1.12.Sdk.linux-x64.7.0.0-dev.nupkg
-rw-r--r-- 1 aleksey aleksey 9.5K Sep  9 13:05 Microsoft.NET.Workload.Emscripten.net6.Manifest-7.0.100.7.0.0-dev.nupkg
-rw-r--r-- 1 aleksey aleksey 9.7K Sep  9 13:05 Microsoft.NET.Workload.Emscripten.net7.Manifest-7.0.100.7.0.0-dev.nupkg
```

So the emscripten SDK is now 196 megs, and the cache sdk is 50 megs

And the Cache SDK contents are listed here https://gist.github.com/lambdageek/53991645623340551e65c4be4213bfa0

**Comment by lambdageek** [comment]
@lewing pushed the workload manifest changes, too.

**Comment by lewing** [comment]
maybe Microsoft.NET.Runtime.Emscripten.Cache.Sdk ->  Microsoft.NET.Runtime.Emscripten.Cache?

File: eng/sdk_files/Emscripten.Cache.targets [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<Project>
+  <Target Name="_EmscriptenCacheSdkOverrideWasmCachePath" BeforeTargets="_PrepareForWasmBuildNative" Condition="'$(WasmBuildNative)' == 'true'">
+    <PropertyGroup>
+      <WasmCachePath Condition="'$(WasmCachePath)' == ''">$([MSBuild]::NormalizeDirectory($(EmscriptenCacheSdkCacheDir)))</WasmCachePath>
```
Comment (lambdageek) [inline eng/sdk_files/Emscripten.Cache.targets:4]: @radical do I need quotes here? how will WasmApp.Native.targets deal with spaces?

Comment (lambdageek) [reply eng/sdk_files/Emscripten.Cache.targets:4]: looks like we don't have quotes for `EmscriptenPythonToolsPath` and that seems to work alright

Comment (radical) [reply eng/sdk_files/Emscripten.Cache.targets:4]: Quote the path where we used the property, instead of here.

**Comment by joeloff** [comment]
> @joeloff if emsdk adds a new pack is that going to cause you issues?

nope, should be fine. thanks for checking

File: eng/emsdk.proj [hunk 2]
```diff
@@ -124,29 +124,42 @@
...
+      <CacheSdkFilesDir>$(ArtifactsObjDir)\.cache-sdk\</CacheSdkFilesDir>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <CacheSdkFilesToMove Include="$(ArtifactsObjDir)upstream\emscripten\cache\**\*" />
+    </ItemGroup>
+
+    <!-- move the emscripten cache to a separate directory -->
+    <Move SourceFiles="@(CacheSdkFilesToMove)"
+	  DestinationFiles="$(CacheSdkFilesDir)emscripten\cache\%(RecursiveDir)%(Filename)%(Extension)" />
```
Comment (radical) [inline eng/emsdk.proj:137]:
```suggestion
          DestinationFiles="$(CacheSdkFilesDir)emscripten\cache\%(RecursiveDir)%(Filename)%(Extension)" />
```

File: eng/nuget/Microsoft.NET.Runtime.Emscripten.Cache/Microsoft.NET.Runtime.Emscripten.Cache.props [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<Project>
```
Comment (radical) [inline eng/nuget/Microsoft.NET.Runtime.Emscripten.Cache/Microsoft.NET.Runtime.Emscripten.Cache.props:1]: Does this need to be a separate file?

File: eng/sdk_files/Emscripten.Cache.props [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<Project>
+  <PropertyGroup>
+    <EmscriptenCacheSdkToolsPath>$(MSBuildThisFileDirectory)..\tools\</EmscriptenCacheSdkToolsPath>
+    <EmscriptenCacheSdkCacheDir>$(EmscriptenCacheSdkToolsPath)emscripten\cache\</EmscriptenCacheSdkCacheDir>
```
Comment (radical) [inline eng/sdk_files/Emscripten.Cache.props:4]: Use `$([MSBuild]::NormalizeDirectory(..))`

File: eng/nuget/Microsoft.NET.Runtime.Emscripten.Cache/Microsoft.NET.Runtime.Emscripten.Cache.props (hunk 3 above)
Comment (lambdageek) [reply eng/nuget/Microsoft.NET.Runtime.Emscripten.Cache/Microsoft.NET.Runtime.Emscripten.Cache.props:1]: Not sure.  I'm copying what the Emscripten.Sdk.pkgproj did

**Comment by lambdageek** [comment]
I ran the workload build tests in dotnet/runtime with the nugets from this PR (by modifying `NuGet.config` and `eng/Versions.{props,Details.xml}` in dotnet/runtme).  All the Blazor tests pass.

I also built a multi-threaded runtime and used it with the workload from this PR to build the multithreaded raytracer demo.  That also worked.

**Review by radical — APPROVED** [review]

**Pull request merged**